Assets/_Dev/Runtime/Spawner.cs

```diff
@@ -8,62 +8,123 @@ namespace Playground
-        private float spawnRate = 1f;
+        [SerializeField, Tooltip("The list of defined waves used for spawning.")]
+        private WaveDefinition[] waves;
         [SerializeField, Tooltip("The radius around the spawner to spawn enemies.")]
         internal float spawnRadius = 5f;
-        [SerializeField, Tooltip("The duration of each spawn wave in seconds.")]
-        private float waveDuration = 10f;
         [SerializeField, Tooltip("The duration of the wait between each spawn wave in seconds.")]
         private float waveWait = 5f;
+        [SerializeField, Tooltip("When defined waves are exhausted, generate more?.")]
+        private bool generateWaves = true;
 
         [Header("Events")]
         [SerializeField, Tooltip("The event to trigger when this spawner is destroyed.")]
         public UnityEvent onDestroyed;
+        [SerializeField, Tooltip("The event to trigger when all waves are complete.")]
+        public UnityEvent onWavesComplete;
 
         List<BasicEnemyController> spawnedEnemies = new List<BasicEnemyController>();
 
+        private int currentWaveIndex = -1;
+
+        private WaveDefinition currentWave;
+
         private void Start()
         {
-            StartCoroutine(SpawnEnemies());
+            StartCoroutine(SpawnWaves());
+        }
+
+        private void NextWave()
+        {
+            currentWaveIndex++;
+            if (currentWaveIndex >= waves.Length)
+            {
+                if (!generateWaves)
+                {
+                    Debug.LogWarning("No more waves to spawn.");
+                    currentWave = null;
+                    onWavesComplete?.Invoke();
+                    StopCoroutine(SpawnWaves());
+                    return;
+                }
+                currentWave = GenerateNewWave();
+                return;
+            }
+            Debug.Log($"Starting wave {currentWaveIndex + 1} of {waves.Length}...");
+            currentWave = waves[currentWaveIndex];
+            currentWave.Reset();
+        }
+
+        private WaveDefinition GenerateNewWave()
+        {
+            Debug.Log("Generating new wave...");
+            WaveDefinition newWave = ScriptableObject.CreateInstance<WaveDefinition>();
+            // populate newWave with random values based loosely on the previous wave
+            var lastWave = currentWave;
+            if (lastWave == null)
+            {
+                lastWave = waves[waves.Length - 1];
+            }
+            // collect all enemy prefabs from all waves and then randomly select from them
+            List<BasicEnemyController> enemyPrefabs = new List<BasicEnemyController>();
+            for (int i = 0; i < waves.Length; i++)
+            {
+                enemyPrefabs.AddRange(waves[i].EnemyPrefabs);
+            }
+            newWave.Init(
+                enemyPrefabs.ToArray(),
+                Mathf.Max(lastWave.SpawnRate + Random.Range(-1f, 1f), 1f), // TODO: make this change more interesting?
```

Comment:
So the wave can be spawning slower than the previous wave, making it easier? Perhaps this should be something like `* Random.Range(0.8f, 1f).

In addition since this is maxing at 1 it possible to have a spawn rate of 0 (1-1)

Reply:
oh hmm good point in that I did that before I reworked how the rate works, changing it to be how frequently in seconds within the duration that it would spawn enemies... so it should be lower to go faster. `Mathf.Max` however should return the highest value, so I was trying to make the minimum be 1

Reply:
Yep, my bad. So no 0 here. Still has the chance of going down, we can work on this in balancing